v2/pkg/engine/resolve/resolve.go

```diff
@@ -285,31 +270,65 @@ type trigger struct {
 }
 
 type sub struct {
-	mux       sync.Mutex
 	resolve   *GraphQLSubscription
+	resolver  *Resolver
+	ctx       *Context
 	writer    SubscriptionResponseWriter
 	id        SubscriptionIdentifier
-	completed chan struct{}
+	heartbeat bool
 	lastWrite time.Time
+	completed chan struct{}
 	// executor is an optional argument that allows us to "schedule" the execution of an update on another thread
 	// e.g. if we're using SSE/Multipart Fetch, we can run the execution on the goroutine of the http request
 	// this ensures that ctx cancellation works properly when a client disconnects
 	executor chan func()
+	// workChan is used to send work to the writer goroutine. All work is processed sequentially.
+	workChan chan func()
 }
 
-func (r *Resolver) executeSubscriptionUpdate(ctx *Context, sub *sub, sharedInput []byte) {
+// startWorker runs in its own goroutine to process fetches and write data to the client synchronously
+// it also takes care of sending heartbeats to the client but only if the subscription supports it
+func (s *sub) startWorker() {
+	heartbeatTicker := time.NewTicker(s.resolver.heartbeatInterval)
```

Comment:
I'd remove the heartbeat if the subscription didn't need one.
This creates additional overhead like a goroutine internally.